server/controllers/caseNotesController.ts

```diff
@@ -20,63 +29,332 @@ export default class CaseNotesController {
+      const prisonerDisplayName = formatName(firstName, undefined, lastName, { style: NameFormatStyle.firstLast })
+
+      // If user cannot view this prisoner's case notes, redirect to 404 page
+      if (!canViewCaseNotes(res.locals.user, { prisonId, restrictedPatient } as Prisoner)) {
+        return res.render('notFound.njk', {
+          url: req.headers.referer || `/prisoner/${prisonerNumber}`,
+        })
+      }
+
+      const now = new Date()
+      const caseNoteFlash = req.flash('caseNote')
+      const { type, subType } = req.query || {}
+      const formValues: CaseNoteForm = caseNoteFlash?.length
+        ? (caseNoteFlash[0] as never)
+        : {
+            type: type as string,
+            subType: subType as string,
+            text: '',
+            date: formatDate(now.toISOString(), 'short'),
+            hours: now.getHours().toString().padStart(2, '0'),
+            minutes: now.getMinutes().toString().padStart(2, '0'),
+          }
+      const errors = req.flash('errors')
+
+      const caseNoteTypes = await this.caseNotesService.getCaseNoteTypesForUser(userToken)
+      const { types, subTypes, typeSubTypeMap } = this.mapCaseNoteTypes(caseNoteTypes, formValues.type, true)
+
+      // Generate back link based on where the user came from - default to profile overview if no referer
+      const addCaseNoteRefererUrlFlash = req.flash('addCaseNoteRefererUrl')
+
+      const refererUrl = addCaseNoteRefererUrlFlash?.length
+        ? addCaseNoteRefererUrlFlash[0]
+        : req.headers.referer || `/prisoner/${prisonerNumber}`
+
+      return res.render('pages/caseNotes/addCaseNote', {
+        today: formatDate(now.toISOString(), 'short'),
+        refererUrl,
+        prisonerDisplayName,
+        prisonerNumber,
+        formValues,
+        types,
+        subTypes,
+        typeSubTypeMap,
+        behaviourPrompts: this.chooseBehaviourPrompts(),
+        errors,
       })
     }
+  }
 
-    let addCaseNoteLinkUrl: string
-    if (canAddCaseNotes(res.locals.user, prisonerData)) {
-      addCaseNoteLinkUrl = `${config.serviceUrls.digitalPrison}/prisoner/${prisonerData.prisonerNumber}/add-case-note`
+  public post(): RequestHandler {
+    return async (req: Request, res: Response) => {
+      const { prisonerNumber } = req.params
+      const { type, subType, text, date, hours, minutes, refererUrl } = req.body
+      const caseNote = {
+        type,
+        subType,
+        text,
+        date,
+        hours,
+        minutes,
+      }
+      const errors = this.validate(type, subType, text, date, hours, minutes)
+      if (!errors.length) {
+        try {
+          await this.caseNotesService.addCaseNote(res.locals.user.token, prisonerNumber, caseNote)
+        } catch (error) {
+          if (error.status === 400) {
+            errors.push({ text: error.message, href: '' })
+          } else throw error
+        }
+      }
```

Comment:
Why is all this being done in the handler? Isn't there middleware to validate forms?

Reply:
Just what I'm used to I guess. Seems like a good suggestion to refactor so will do that.

Reply:
Well I think when I was here last it was standard, there are certainly examples in other repos. Maybe not the standard approach now?

Reply:
Perhaps it's a new ticket to implement a reusable form validation mechanism. We certainly had a decent example in Managing Intelligence that I'm sure I could dig out

Reply:
No worries... I've done the refactor, just finishing off the unit tests for the new approach.